File: demos/tebd.py
import numpy as np
from scipy.linalg import expm
from scipy import integrate


def itebd(G_list, l_list, U, chi_max):
    " Updates the G and s matrices using U and the TEBD protocol "
    d = G_list[0].shape[0]

    for ibond in [0, 1]:
        ia = np.mod(ibond, 2)
        ib = np.mod(ibond + 1, 2)

        chi1 = G_list[ia].shape[1]
        chi3 = G_list[ib].shape[2]

        # Construct theta
        theta = np.tensordot(np.diag(l_list[ib]), G_list[ia], axes=(1, 1))
        theta = np.tensordot(theta, np.diag(l_list[ia], 0), axes=(2, 0))
        theta = np.tensordot(theta, G_list[ib], axes=(2, 1))
        theta = np.tensordot(theta, np.diag(l_list[ib], 0), axes=(3, 0))

        # Apply U
        theta = np.tensordot(theta, np.reshape(U, (d, d, d, d)), axes=([1, 2], [0, 1]))

        # SVD
        theta = np.reshape(np.transpose(theta, (2, 0, 3, 1)), (d * chi1, d * chi3))  # ip a jp b
        X, Y, Z = np.linalg.svd(theta)
        Z = Z.T
        chi2 = np.min([np.sum(Y > 10. ** (-10)), chi_max])

        # Truncate
        l_list[ia] = Y[0:chi2] / np.sqrt(sum(Y[0:chi2] ** 2))

        X = np.reshape(X[:, 0:chi2], (d, chi1, chi2))
        G_list[ia] = np.transpose(np.tensordot(np.diag(l_list[ib] ** (-1)), X, axes=(1, 1)), (1, 0, 2))

        Z = np.transpose(np.reshape(Z[:, 0:chi2], (d, chi3, chi2)), (0, 2, 1))
        G_list[ib] = np.tensordot(Z, np.diag(l_list[ib] ** (-1)), axes=(2, 0))


def site_expectation_value(G_list, l_list, O):
    " Expectation value for a site operator "
    E = []
    for isite in range(0, 2):
        theta = np.tensordot(np.diag(l_list[np.mod(isite - 1, 2)]), G_list[isite], axes=(1, 1))
        theta = np.tensordot(theta, np.diag(l_list[isite]), axes=(2, 0))
        theta_O = np.tensordot(theta, O, axes=(1, 0)).conj()
        E.append(np.squeeze(np.tensordot(theta_O, theta, axes=([0, 1, 2], [0, 2, 1]))).item())
    return (E)


def bond_expectation_value(G_list, l_list, O):
    " Expectation value for a site operator "
    E = []
    for ibond in range(0, 2):
        ia = np.mod(ibond, 2)
        ib = np.mod(ibond + 1, 2)
        theta = np.tensordot(np.diag(l_list[ib]), G_list[ia], axes=(1, 1))
        theta = np.tensordot(theta, np.diag(l_list[ia], 0), axes=(2, 0))
        theta = np.tensordot(theta, G_list[ib], axes=(2, 1))
        theta = np.tensordot(theta, np.diag(l_list[ib], 0), axes=(3, 0))
        theta_O = np.tensordot(theta, np.reshape(O, (d, d, d, d)), axes=([1, 2], [0, 1])).conj()
        E.append(np.squeeze(np.tensordot(theta_O, theta, axes=([0, 1, 2, 3], [0, 3, 1, 2]))).item())

    return (E)


######## Define the simulation parameter ######################
chi_max = 10
delta = 0.01
N = 1000;
d = 2;
g = 0.5

########### Define Ising Hamiltonian and get U ################
sx = np.array([[0., 1.], [1., 0.]])
sz = np.array([[1., 0.], [0., -1.]])

H = -np.kron(sz, sz) + g * np.kron(sx, np.eye(2, 2))
U = expm(-delta * H)

############### Initial state : |0000> ########################
Ga = np.zeros((d, 1, 1), dtype=float)
Ga[0, 0, 0] = 1.
Gb = np.zeros((d, 1, 1), dtype=float)
Gb[0, 0, 0] = 1.
G_list = [Ga, Gb]

la = np.zeros(1)
la[0] = 1.
lb = np.zeros(1)
lb[0] = 1.
l_list = [la, lb]

for step in range(1, N):
    itebd(G_list, l_list, U, chi_max)
print("m       = ", np.mean(site_expectation_value(G_list, l_list, sz)))
print("E_itebd = ", np.mean(bond_expectation_value(G_list, l_list, H)))

############### Get the exact energy #########################
f = lambda k, g: -2 * np.sqrt(1 + g ** 2 - 2 * g * np.cos(k)) / np.pi / 2.
E0_exact = integrate.quad(f, 0, np.pi, args=(g,))[0]
print("E_exact = ", E0_exact)

File: demos/tebd.jl
using TensorOperations
using QuadGK
using LinearAlgebra
using LinearAlgebra.LAPACK
using Einsum
using Statistics

function itebd(G_list, l_list, U, chi)
    d = size(G_list[1])[1]

    loop = 0
    for ibond in 1:2
        loop +=1
        # println("Loop number: ", loop)
        # println("Before loop G list: ", G_list)
        # println(ibond)
        ia = ibond
        ib = mod(ibond, 2) + 1
        # println("ia = ", ia, "\nib = ", ib)

        chi1 = size(G_list[ia])[2]
        chi3 = size(G_list[ib])[3]
        # println(chi1, chi3)

        # Construct theta
        Gia = G_list[ia]
        Gib = G_list[ib]
        # println("GIA ", Gia)
        # println("l list = ", l_list)
        Dia = diagm(0=>l_list[ia])
        Dib = diagm(0=>l_list[ib])
        # println("Dia: ", Dia)
        # theta0 = zeros((size(Dib, 1), size(Gia, 1), size(Gib, 1), size(Dib, 2)))
        # theta  = zeros((size(Dib, 1), size(Dib, 2), size(Gia, 1), size(Gib, 1)))
        # println("GList ia",Gia)
        # println("GList ib",Gib)
        # println("DList ia",Dia)
        # println("DList ib",Dib)
        
        # @tensor theta0[i, j, k, l] := Dib[i, b] * Gia[j, b, d] * Dia[d, e] * Gib[k, e, g] * Dib[g, l]
        # println("theta0 = ", theta0)
        # println("theta = ", theta)
        # println("Size of theta0 = ", size(theta0))

        # Apply U 
        # tempU = reshape(U, (d,d,d,d))
        # println("Size of temp U = ", size(tempU))
        # println("Size of theta = ", size(theta))
        # @tensor theta[i, j, k, l] := theta0[i, a, b, j] * tempU[a, b, k, l]
        
        """Multiple steps"""
        # @tensor theta[i, j, k] := Dib[i, a] * Gia[j, a, k]
        # # println("theta[i, j, k] := Dib[i, a] * Gia[j, a, k]\n", theta)
        # @tensor theta[i, j, k] := theta[i, j, a] * Dia[a, k]
        # # println("theta[i, j, k] := theta[i, j, a] * Dia[a, k]\n", theta)
        # @tensor theta[i, j, k, l] := theta[i, j, a] * Gib[k, a, l]
        # # println("theta[i, j, k, l] := theta[i, j, a] * Gib[k, a, l]\n", theta)
        # @tensor theta[i, j, k, l] := theta[i, j, k, a] * Dib[a, l]
        # # println("theta[i, j, k, l] := theta[i, j, k, a] * Dib[a, l]\n", theta)
        # Ur = reshape(U, (d,d,d,d))

        # # theta = permutedims(theta, (4, 3, 2, 1))
        # Ur = permutedims(Ur, (4, 3, 2, 1))
        # # println("Ur = ", Ur)
        # @tensor theta[i, j, k, l] := theta[i, a1, a2, j] * Ur[a1, a2, k, l]

        """One step"""
        @tensor theta[i, j, k, l] := Dib[i, b] * Gia[a, b, h] * Dia[h, e] * Gib[f, e, g] * Dib[g, j] * reshape(U, (d,d,d,d))[a, f, k, l]

        # SVD
        theta = reshape(permutedims(theta, (3, 1, 4, 2)), (d * chi1, d * chi3))
        
        # println("Theta reshaped = ", theta, " of the type", typeof(theta))
        # println("The size of theta is ", size(theta))
        # F = svd(theta)
        # X = F.U
        # Y = F.S
        # Z = F.Vt
        (X, Y, Z) = LAPACK.gesdd!('A', Matrix(theta))
        # println("X=",X)
        # println("Y=",Y)
        # println("Z=",Z)
        Z = permutedims(Z)
        notzero = Y.> (10^-10)
        # println("notezero=",notzero)
        # println("Sum = ", sum(notzero))
        chi2 = minimum([sum(notzero), chi_max])

        # Truncate
        l_list[ia] = Y[1:chi2] / sqrt(sum(Y[1:chi2] .^ 2))

        X = reshape(X[:, 1:chi2], (d, chi1, chi2))

        Dibinv = diagm(0=>l_list[ib] .^ (-1))
        @tensor temp[i, j, k]:=Dibinv[i, a]*X[j, a, k]
        G_list[ia] = permutedims(temp, (2, 1, 3))

        Z = permutedims(reshape(Z[:, 1:chi2], (d, chi3, chi2)), (1, 3, 2))
        temp2 = zeros(size(Z, 1), size(Z, 2), size(Dibinv, 2))
        @tensor temp2[i, j, k]=Z[i, j, a]*Dibinv[a, k]
        G_list[ib] = temp2

        # println("After loop G list: ", G_list)
        # println(" ")
        

    end
end

function site_expectation_value(G_list, l_list, O)
    # Expectation value for a site operator
    E = []
    for isite in 1:2
        Gisite = G_list[isite]
        Disite = diagm(0=>l_list[mod(isite, 2) + 1])
        Disite2= diagm(0=>l_list[isite])
        # theta  = zeros(size(Disite, 1), size(Gisite, 1), size(Disite2, 2))
        @tensor theta[i, j, k] := Disite[i, a]*Gisite[j, a, b]*Disite2[b, k]

        # theta_O= zeros(size(theta, 1), size(theta, 3), size(O, 2))
        @tensor theta_O[i, j, k] := theta[i, a, j] * O[a, k]
        conj!(theta_O)

        @tensor E_cal[]:=theta_O[i, j, k] * theta[i, k, j]
        push!(E, E_cal)
    end

    # println("E = ", E)
    
    return E
    
end

function bond_expectation_value(G_list, l_list, O)
    # Expectation value for a site operator
    E = []
    for ibond in 1:2
        ia = ibond
        ib = mod(ibond, 2)+1

        Gia= G_list[ia]
        Gib= G_list[ib]
        Dia= diagm(0=>l_list[ia])
        Dib= diagm(0=>l_list[ib])
        
        @tensor theta[i, j, k, l]:=Dib[i, a]*Gia[j, a, b]*Dia[b, d]*Gib[k, d, f]*Dib[f, l]
        O2 = reshape(O, (d,d,d,d))
        @tensor theta_O[i,j,k,l]:=theta[i, a1, a2, j]*O2[a1, a2, k, l]
        conj!(theta_O)
        @tensor E_cal[]:=theta_O[a0, a1, a2, a3]*theta[a0, a2, a3, a1]
        push!(E, E_cal)
    end

    return E
end

t0 = time()
# Define the simulation parameters
chi_max   = 10
δ       = 0.01
N       = 200
d       = 2
g       = 0.5

# Define Ising Hamiltonian and get using
id      = diagm(0=>ones(2))
sx      = [ [0.     1.  ]
            [1.     0.  ]]
sz      = [ [1.     0.  ]
            [0.    -1.  ]]

H       = -1 * kron(sz, sz) + g * kron(sx, id)
U       = exp(- δ * H )
# println("U=",U)
# Initial state: |0000>
Ga = zeros(Float64, (d, 1, 1))
Ga[1, 1, 1] = 1.
Gb = zeros(Float64, (d, 1, 1))
Gb[1, 1, 1] = 1.
G_list = [Ga, Gb]

la = zeros(1)
la[1] = 1.
lb = zeros(1)
lb[1] = 1.
l_list = [la, lb]

for step in 2:N
    @eval begin
    itebd(G_list, l_list, U, chi_max)
    end
end

println("m       = ", mean(site_expectation_value(G_list, l_list, sz)))
println("E_itebd = ", mean(bond_expectation_value(G_list, l_list, H)))

# Get the exact energy

function ff(k)
    return -2 * sqrt(1 + g * g - 2 * g * cos(k)) / pi / 2.
end
E0_exact = quadgk(ff, 0, pi)
println("E_exact = ", E0_exact)
# println("PROGRAMME FINISHED!!!\n\n\n")
t1 = time()
println("Time elapsed = ", t1-t0)
